File: Validation.py
import os.path

class Validation:

    @staticmethod
    def index_check(func):
        def inner(lst, filename, index):
            index = input("Index: ")
            while not Validation().digit_check(index):
                index = input("Index must be a positive number: ")
            index = int(index)
            while index != 0 and lst.length() == 0:
                index = input("Your list is empty, you can choose only \"0\" index: ")
                while not Validation().digit_check(index):
                    index = input("Index must be a positive number: ")
                index = int(index)
            while index > lst.length() or index < 0:
                index = input("Index must be in list range: ")
                while not Validation().digit_check(index):
                    index = input("Index must be a positive number: ")
                index = int(index)
            func(lst,filename, index)
        return inner

    @staticmethod
    def file_check(func):
        def inner(lst, filename, index):
            filename = input("Input name of data file: ")
            while not os.path.exists(filename):
                filename = input("File doesn't exist: ")
            func(lst, filename, index)
        return inner

    @staticmethod
    def limit_check(func):
        def inner(lst, limit, index):
            limit = input("Input limit for iterator: ")
            while not Validation().digit_check(limit):
                limit = input("Limit must be a positive number: ")
            func(lst, limit, index)
        return inner


    @staticmethod
    def digit_check(var):
        return var.isdigit()

    @staticmethod
    def positive_check(var):
        minus_count = 0
        dot_count = 0
        for i in var:
            if i == "-":    minus_count+=1
            if i == ".":    dot_count+=1
        wo_minusdots = ""
        if minus_count <=1:
            for i in var:
                if i == "-" or i == ".": continue
                wo_minusdots+=i
            for i in wo_minusdots:
                if not i.isdigit():
                    return False
            if minus_count > 0:
                return var[0] == "-"
            dot_position = True
            if dot_count > 0:
                dot_position = (var[0]!="." and var[len(var)-1]!=".")
            result = Validation().digit_check(wo_minusdots) and dot_count<=1 and dot_position
        else: return False
        return result

File: Validation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.IO;
using System.Globalization;
using System.Reflection;
using Newtonsoft.Json;


namespace Task1
{
    class Validation
    {
        public static bool CheckPath(string path)
        {
            if (!File.Exists(@path))
                Console.WriteLine("File doesn't exist!");
            return File.Exists(@path);
        }
        public static bool IsNumber(string value)
        {
            bool result = false;
            ulong num = 0;
            if (value.Contains(","))
                result = value.Count(c => c == ',') == 1 && ulong.TryParse(value.Replace(",", ""), out num);
            else
                result = ulong.TryParse(value, out num) && true;
            if (!result)
                Console.WriteLine($"{value} doesn't look like a number!");
            return result;
        }
        public static bool IsHolder(string value)
        {
            if (!value.Replace(" ", "").All(Char.IsLetter))
                Console.WriteLine($"{value} must consist only letters!");
            if (!(1 == value.Count(c => c == ' ')))
                Console.WriteLine($"Invalid SPACE count in {value}!");
            return value.Replace(" ", "").All(Char.IsLetter) && 1 == value.Count(c => c == ' ');
        }
        public static bool IsCurrency(string value, Type enumType)
        {
            return Enum.GetNames(enumType).Contains(value.ToLower());
        }
        public static bool IsIban(string value)
        {
            if (!(value.StartsWith("UA")))
                Console.WriteLine($"{value} must starts with UA!");
            if (!(3 == value.Count(c => c == ' ')))
                Console.WriteLine($"Invalid SPACE count in {value}!");
            return value.StartsWith("UA") && 3 == value.Count(c => c == ' ') && IsNumber(value.Replace(" ", "").Replace("UA", "")) && value.Length == 19;
        }
        public static bool IsDate(string value)
        {
            DateTime temp;
            if (value.Length > 10)
                value = value.Remove(10);
            if (!(DateTime.TryParseExact(value, "dd.MM.yyyy", null, DateTimeStyles.None, out temp) && (temp <= DateTime.Now.Date)))
                Console.WriteLine($"{value} must have dd.MM.yyyy format and <= today!");
            return DateTime.TryParseExact(value, "dd.MM.yyyy", null, DateTimeStyles.None, out temp) && (temp <= DateTime.Now.Date);
        }
        public static bool IsValidObject<T>(string json_obj)
        {
            bool result = true;
            try
            {
                T temp = JsonConvert.DeserializeObject<T>(json_obj);
                PropertyInfo[] props = typeof(T).GetProperties(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
                for (int i = 0; i < props.Length; i++)
                {
                    if (!result)
                    {
                        Console.WriteLine($"{props[i - 1].GetValue(temp)} is invalid!");
                        return false;
                    }
                    if (props[i].PropertyType.IsEnum)
                        result = IsCurrency(props[i].GetValue(temp).ToString(), props[i].PropertyType);
                    if (props[i].PropertyType == typeof(int) || props[i].PropertyType == typeof(double))
                        result = IsNumber(props[i].GetValue(temp).ToString()) && true;
                    if (props[i].Name.ToLower().Contains("holder"))
                        result = IsHolder(props[i].GetValue(temp).ToString()) && true;
                    if (props[i].Name.ToLower().Contains("iban"))
                        result = IsIban(props[i].GetValue(temp).ToString()) && true;
                    if (props[i].Name.ToLower().Contains("date"))
                        result = IsDate(props[i].GetValue(temp).ToString()) && true;
                }
            }
            catch
            {
                Console.WriteLine($"Object below was skipped due to incorrect data!\n{json_obj}");
                Console.Write("\nMake sure:\n\tid: integer>0\n\tamount: num>0\n\tfee_amount: num>0\n\tcurency: eur,usd,uah\n\tiban: UA11 1111 1111 1111\n\tholder: Name Surname\n\tdate: yyyy.MM.dd( <= today)\n");
                result = false;
                return result;
            }
            return result;
        }
    }
}
